cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/preview/PreviewDatasetFramework.java

```diff
@@ -56,97 +92,138 @@ public PreviewDatasetFramework(DatasetFramework local, DatasetFramework actual,
 
   @Override
   public void addModule(DatasetModuleId moduleId, DatasetModule module) throws DatasetManagementException {
+    localDatasetFramework.addModule(moduleId, module);
   }
 
   @Override
   public void addModule(DatasetModuleId moduleId, DatasetModule module, Location jarLocation)
     throws DatasetManagementException {
+    // called while deploying the new application from DatasetModuleDeployer stage.
+    // any new module should be deployed in the preview space.
+    localDatasetFramework.addModule(moduleId, module, jarLocation);
   }
 
   @Override
   public void deleteModule(DatasetModuleId moduleId) throws DatasetManagementException {
+    localDatasetFramework.deleteModule(moduleId);
   }
 
   @Override
   public void deleteAllModules(NamespaceId namespaceId) throws DatasetManagementException {
+    localDatasetFramework.deleteAllModules(namespaceId);
   }
 
   @Override
   public void addInstance(String datasetTypeName, DatasetId datasetInstanceId,
                           DatasetProperties props) throws DatasetManagementException, IOException {
+    localDatasetFramework.addInstance(datasetTypeName, datasetInstanceId, props);
   }
 
   @Override
   public void updateInstance(DatasetId datasetInstanceId,
                              DatasetProperties props) throws DatasetManagementException, IOException {
+    // allow updates to the datasets in preview space only
+    if (localDatasetFramework.hasInstance(datasetInstanceId)) {
+      localDatasetFramework.updateInstance(datasetInstanceId, props);
+    }
   }
 
   @Override
   public Collection<DatasetSpecificationSummary> getInstances(NamespaceId namespaceId)
     throws DatasetManagementException {
-    return null;
+    return localDatasetFramework.getInstances(namespaceId);
   }
 
   @Nullable
   @Override
   public DatasetSpecification getDatasetSpec(DatasetId datasetInstanceId) throws DatasetManagementException {
-    return null;
+    if (datasetNames.contains(datasetInstanceId.getDataset())) {
+      return actualDatasetFramework.getDatasetSpec(datasetInstanceId);
+    }
+    return localDatasetFramework.getDatasetSpec(datasetInstanceId);
   }
 
   @Override
   public boolean hasInstance(DatasetId datasetInstanceId) throws DatasetManagementException {
-    return false;
+    if (datasetNames.contains(datasetInstanceId.getDataset())) {
+      return actualDatasetFramework.hasInstance(datasetInstanceId);
```

Comment:
Since we are using the base dataset framework, we should enforce authorization here (see how the `LineageWriterDatasetFramework` does it through the `DefaultDatasetRuntimeContext.execute` method).